Wrappers/Python/cil/optimisation/algorithms/SPDHG.py

```diff
@@ -42,202 +46,460 @@ class SPDHG(Algorithm):
+        super(SPDHG, self).__init__(
+            update_objective_interval=update_objective_interval)
 
+        self.set_up(f=f, g=g, operator=operator, sigma=sigma, tau=tau,
+                    initial=initial,  sampler=sampler, prob_weights=prob_weights,  **kwargs)
 
-    def set_up(self, f, g, operator, tau=None, sigma=None, \
-               initial=None, prob=None, gamma=1., norms=None):
-
+    def set_up(self, f, g, operator, sigma=None, tau=None,
+               initial=None,   sampler=None, prob_weights=None, **deprecated_kwargs):
         '''set-up of the algorithm
-        Parameters
-        ----------
-        f : BlockFunction
-            Each must be a convex function with a "simple" proximal method of its conjugate
-        g : Function
-            A convex function with a "simple" proximal
-        operator : BlockOperator
-            BlockOperator must contain Linear Operators
-        tau : positive float, optional, default=None
-            Step size parameter for Primal problem
-        sigma : list of positive float, optional, default=None
-            List of Step size parameters for Dual problem
-        initial : DataContainer, optional, default=None
-            Initial point for the SPDHG algorithm
-        prob : list of floats, optional, default=None
-            List of probabilities. If None each subset will have probability = 1/number of subsets
-        gamma : float
-            parameter controlling the trade-off between the primal and dual step sizes
-
-        **kwargs:
-        norms : list of floats
-            precalculated list of norms of the operators
         '''
         log.info("%s setting up", self.__class__.__name__)
+
         # algorithmic parameters
         self.f = f
         self.g = g
         self.operator = operator
-        self.tau = tau
-        self.sigma = sigma
-        self.prob = prob
-        self.ndual_subsets = len(self.operator)
-        self.gamma = gamma
-        self.rho = .99
-
-        if self.prob is None:
-            self.prob = [1/self.ndual_subsets] * self.ndual_subsets
-
-
-        if self.sigma is None:
-            if norms is None:
-                # Compute norm of each sub-operator
-                norms = [operator.get_item(i,0).norm() for i in range(self.ndual_subsets)]
-            self.norms = norms
-            self.sigma = [self.gamma * self.rho / ni for ni in norms]
-        if self.tau is None:
-            self.tau = min( [ pi / ( si * ni**2 ) for pi, ni, si in zip(self.prob, norms, self.sigma)] )
-            self.tau *= (self.rho / self.gamma)
+
+        if not isinstance(operator, BlockOperator):
+            raise TypeError("operator should be a BlockOperator")
+
+        self._ndual_subsets = len(self.operator)
+        
+
+        # Set up sampler and prob weights from deprecated "prob" argument
+        sampler = self._deprecated_set_prob(deprecated_kwargs, prob_weights, sampler) 
```

Comment:
this isn't very clear. I can see if I pass `sampler=None` and `prob` a sampler is created. Could this be handled directly by the if statement in ln 167 that creates a sampler?

Otherwise the naming implies it's just about setting prob, which I think would be sufficient. Then if Sampler is `None` it gets created as planned with  `self._prob_weights`

I can see you're trying to split the deprecated set up from the main code, but it's confusing that it's creating the sampler in 2 places.

Reply:
I think sorted